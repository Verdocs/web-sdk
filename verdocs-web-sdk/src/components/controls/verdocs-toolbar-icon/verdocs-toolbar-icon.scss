@import '../../../theme.scss';

verdocs-toolbar-icon {
  font-family: $verdocs-primary-font;
  //display: inline-block;
  display: flex;
  position: relative;
  align-items: center;
  justify-content: center;

  .icon {
    display: inline-block;
    cursor: pointer;

    svg {
      fill: $choice-available-color;
    }
  }

  .tooltip {
    display: none;
    box-shadow: 0 0 10px 1px #999999;
    background: #ffffff;
    color: $label-color;
    max-width: 240px;
    font-weight: bold;
    padding: 5px 10px;
    font-size: 13px;
    position: relative;
    border-radius: 4px;
    z-index: 20000;
    white-space: nowrap;

    .arrow,
    .arrow::before {
      position: absolute;
      width: 8px;
      height: 8px;
      background: inherit;
    }

    .arrow {
      visibility: hidden;
      top: -4px;
    }

    .arrow::before {
      visibility: visible;
      content: '';
      transform: rotate(45deg);
    }

    &[data-show] {
      display: block;
    }

    &[data-popper-placement^='top'] > .arrow {
      bottom: -4px;
    }

    &[data-popper-placement^='bottom'] > .arrow {
      top: -4px;
    }

    &[data-popper-placement^='left'] > .arrow {
      right: -4px;
    }

    &[data-popper-placement^='right'] > .arrow {
      left: -4px;
    }

    // For the floating menu
    &[data-popper-placement^='left'] {
      //margin-right: 22px !important;
      padding: 3px 10px;

      & > .arrow {
        right: -1px;
      }
    }
  }
}
