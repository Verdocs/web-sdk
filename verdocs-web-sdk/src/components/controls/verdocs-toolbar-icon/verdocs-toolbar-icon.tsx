import {createPopper, Instance} from '@popperjs/core';
import {Component, h, Host, Prop, State} from '@stencil/core';
// import {Component, Event, EventEmitter, h, Host, Prop, State} from '@stencil/core';

/**
 * Displays a simple help icon. Upon hover or focus, a tooltip will be displayed with help text.
 */
@Component({
  tag: 'verdocs-toolbar-icon',
  styleUrl: 'verdocs-toolbar-icon.scss',
  shadow: false,
})
export class VerdocsToolbarIcon {
  iconEl: HTMLDivElement;
  tooltip: HTMLDivElement;
  popperInstance: Instance;

  /**
   * Help text to display on hover/focus
   */
  @Prop() text: string = '';

  /**
   * SVG icon to display
   */
  @Prop() icon: string = '';

  /**
   * Triggered when the icon is pressed
   */
  // @Event({composed: true}) press: EventEmitter;

  @State() containerId = `verdocs-toolbar-icon-${Math.random().toString(36).substring(2, 11)}`;

  componentDidLoad() {
    this.popperInstance = createPopper(this.iconEl, this.tooltip, {
      // placement: 'top-end',
      // modifiers: [{name: 'offset', options: {offset: [0, 10]}}],
    });
  }

  disconnectedCallback() {
    if (this.popperInstance) {
      this.popperInstance.destroy();
      this.popperInstance = null;
    }
  }

  show() {
    this.tooltip?.setAttribute('data-show', '');
    this.popperInstance?.update().catch(() => {});
  }

  hide() {
    this.tooltip?.removeAttribute('data-show');
  }

  // handlePress() {
  //   this.press?.emit();
  // }

  render() {
    return (
      <Host class={{}}>
        <div
          aria-describedby={this.containerId}
          class="icon"
          innerHTML={this.icon}
          onMouseEnter={() => this.show()}
          onFocus={() => this.show()}
          onMouseLeave={() => this.hide()}
          onBlur={() => this.hide()}
          ref={el => (this.iconEl = el as HTMLDivElement)}
        />
        <div id={this.containerId} role="tooltip" class="tooltip" data-popper-placement="bottom" ref={el => (this.tooltip = el as HTMLDivElement)}>
          {this.text}
          <div data-popper-arrow="true" class="arrow" />
        </div>
      </Host>
    );
  }
}
