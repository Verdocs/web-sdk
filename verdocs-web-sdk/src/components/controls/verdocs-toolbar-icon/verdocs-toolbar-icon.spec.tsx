import { newSpecPage } from '@stencil/core/testing';
import { VerdocsToolbarIcon } from './verdocs-toolbar-icon';
import {h} from '@stencil/core';

describe('verdocs-toolbar-icon', () => {
   beforeAll(() => {
    jest.spyOn(Math, 'random').mockReturnValue(0.123456789);
  });

  afterAll(() => {
    // restore the original behavior after the tests
    (Math.random as jest.Mock).mockRestore();
  });
  it('renders with default props', async () => {
    const page = await newSpecPage({
      components: [VerdocsToolbarIcon],
      html: `<verdocs-toolbar-icon></verdocs-toolbar-icon>`,
    });
    expect(page.root).toMatchSnapshot();
  });

  it('renders with text, icon, and placement', async () => {
    const icon = '<svg></svg>';
    const page = await newSpecPage({
      components: [VerdocsToolbarIcon],
      template: () => <verdocs-toolbar-icon text="Tooltip" icon={icon} placement="top"></verdocs-toolbar-icon>,
    });
    expect(page.root).toMatchSnapshot();
    expect(page.rootInstance.text).toBe('Tooltip');
    expect(page.rootInstance.icon).toBe(icon);
    expect(page.rootInstance.placement).toBe('top');
  });
});
