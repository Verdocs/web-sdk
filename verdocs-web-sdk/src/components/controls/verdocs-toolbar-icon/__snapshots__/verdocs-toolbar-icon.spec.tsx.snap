// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`verdocs-toolbar-icon renders with default props 1`] = `
<verdocs-toolbar-icon>
  <div aria-describedby="verdocs-toolbar-icon-4fzzzxjyl" class="icon"></div>
  <div class="tooltip" data-popper-placement="bottom" id="verdocs-toolbar-icon-4fzzzxjyl" role="tooltip" style="position: absolute; left: 0; top: 0; margin: 0;">
    <div class="arrow" data-popper-arrow="true" style="position: absolute;"></div>
  </div>
</verdocs-toolbar-icon>
`;

exports[`verdocs-toolbar-icon renders with text, icon, and placement 1`] = `
<verdocs-toolbar-icon>
  <div aria-describedby="verdocs-toolbar-icon-4fzzzxjyl" class="icon">
    <svg></svg>
  </div>
  <div class="tooltip" data-popper-placement="top" id="verdocs-toolbar-icon-4fzzzxjyl" role="tooltip" style="position: absolute; left: 0; top: 0; margin: 0;">
    Tooltip
    <div class="arrow" data-popper-arrow="true" style="position: absolute;"></div>
  </div>
</verdocs-toolbar-icon>
`;
